datafusion/physical-expr/src/aggregate/count_distinct.rs

```diff
@@ -192,6 +262,164 @@ impl Accumulator for DistinctCountAccumulator {
     }
 }
 
+#[derive(Debug)]
+struct NativeDistinctCountAccumulator<T>
+where
+    T: ArrowPrimitiveType + Send,
+    T::Native: Eq + Hash,
+{
+    values: HashSet<T::Native, RandomState>,
+}
+
+impl<T> NativeDistinctCountAccumulator<T>
+where
+    T: ArrowPrimitiveType + Send,
+    T::Native: Eq + Hash,
+{
+    fn new() -> Self {
+        Self {
+            values: HashSet::default(),
+        }
+    }
+}
+
+impl<T> Accumulator for NativeDistinctCountAccumulator<T>
+where
+    T: ArrowPrimitiveType + Send + Debug,
+    T::Native: Eq + Hash,
+{
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        let arr = Arc::new(PrimitiveArray::<T>::from_iter_values(
+            self.values.iter().cloned(),
+        )) as ArrayRef;
+        let list = Arc::new(array_into_list_array(arr)) as ArrayRef;
+        Ok(vec![ScalarValue::List(list)])
+    }
+
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        if values.is_empty() {
+            return Ok(());
+        }
+
+        let arr = as_primitive_array::<T>(&values[0])?;
+        arr.iter().for_each(|value| {
+            if let Some(value) = value {
+                self.values.insert(value);
+            }
+        });
+
+        Ok(())
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        if states.is_empty() {
+            return Ok(());
+        }
+        assert_eq!(
+            states.len(),
+            1,
+            "count_distinct states must be single array"
+        );
+
+        let arr = as_list_array(&states[0])?;
+        arr.iter().try_for_each(|maybe_list| {
+            if let Some(list) = maybe_list {
+                let list = as_primitive_array::<T>(&list)?;
+                self.values.extend(list.values())
+            };
+            Ok(())
+        })
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        Ok(ScalarValue::Int64(Some(self.values.len() as i64)))
+    }
+
+    fn size(&self) -> usize {
+        std::mem::size_of_val(self)
+            + std::mem::size_of_val(&self.values)
+            + (std::mem::size_of::<T::Native>() * self.values.capacity())
```

Comment:
This needs to be adapted for `HashSet`

Reply:
It seems like maybe we just need to add overhead of the size of a hash for each entry? I poked around in the HashSet / HashTable implementation and it is not immeditately clear how much additional overhead there is

Maybe something like
```rust
            + ((std::mem::size_of::<T::Native>() + std::mem::size_of<u64>)) * self.values.capacity())
```

Reply:
There is also some overallocation happening (not sure how much, it might be roughly 1.2x) to reduce collisions.
Hashes are not stored in HashSet (by default), so we shouldn't need to take that into account.

Reply:
Nvm, there is some more complicated allocation pattern happening here:

https://docs.rs/hashbrown/latest/src/hashbrown/raw/mod.rs.html#229

So, roughly 1.125x for larger tables.

Reply:
We probably need to address in different places where we use HashMap / HashSets / RawTable, so I'm fine with keeping this out of this PR for now.

Reply:
My bad, shoud have checked it -- as it uses hashbrown::hashset (which is actually hasmap) maybe we could use same calculation as in hash join inner size estimation (see the [link](https://github.com/apache/arrow-datafusion/blob/819d3577872a082f2aea7a68ae83d68534049662/datafusion/physical-plan/src/joins/hash_join.rs#L748))

The only difference should be
```
let estimated_hastable_size =
        std::mem::size_of::<T::Native> * estimated_buckets + estimated_buckets + size_of::<JoinHashMap>();
```
size of native type, instead of 16 bytes for tuple in join case.

Reply:
Ah, forgot that we do it already in join, should be fine to do it here as well then :)

Reply:
Ported hashtable size estimation from join operator.